File: packages/lingxia-elements/src/inline-native/video.ts
import type { NodeRef } from "./types.js";

export const VIDEO_COMMANDS = [
  "play",
  "pause",
  "stop",
  "seek",
  "enterFullscreen",
  "exitFullscreen",
  "setStreamSource",
] as const;

export type VideoCommandName = (typeof VIDEO_COMMANDS)[number];

export type VideoCommand =
  | { name: "play" }
  | { name: "pause" }
  | { name: "stop" }
  | { name: "seek"; seconds: number }
  | { name: "enterFullscreen" }
  | { name: "exitFullscreen" }
  | { name: "setStreamSource"; options: Record<string, unknown> };

export interface VideoCommandRequest {
  action: "video.command";
  owner: NodeRef;
  requestId: string;
  command: VideoCommand;
}

export interface VideoControlDescriptorBase {
  controlId: string;
  label: string;
  frame: { x: number; y: number; width: number; height: number };
  visible: boolean;
  disabled?: boolean;
}

export type VideoControlDescriptor =
  | (VideoControlDescriptorBase & {
      role: "button";
      pressed?: boolean;
      expanded?: boolean;
      actions: ReadonlyArray<"focus" | "blur" | "activate">;
    })
  | (VideoControlDescriptorBase & {
      role: "slider";
      value: number;
      min: number;
      max: number;
      actions: ReadonlyArray<"focus" | "blur" | "increment" | "decrement" | "setValue">;
    });

export interface VideoControlsSemanticSnapshot {
  action: "video.controlsSemanticSnapshot";
  owner: NodeRef;
  revision: number;
  controls: readonly VideoControlDescriptor[];
}

export function buildVideoCommandRequest(
  owner: NodeRef,
  command: VideoCommand,
  requestId: string
): VideoCommandRequest {
  return { action: "video.command", owner, requestId, command };
}

export function videoCommandUrls(command: VideoCommand): string[] {
  if (command.name !== "setStreamSource") return [];
  const options = command.options ?? {};
  const urls: string[] = [];
  for (const key of ["url", "src", "uri"]) {
    const value = options[key];
    if (typeof value === "string" && value) urls.push(value);
  }
  return urls;
}

export function collectVideoResourceUrls(props: Record<string, unknown>): string[] {
  const urls: string[] = [];
  for (const key of ["src", "poster"]) {
    const value = props[key];
    if (typeof value === "string" && value) urls.push(value);
  }
  const watermark = props.watermark as { resource?: { url?: string } } | undefined;
  if (watermark?.resource && typeof watermark.resource === "object") {
    const url = (watermark.resource as { url?: string }).url;
    if (typeof url === "string" && url) urls.push(url);
  }
  const qualities = props.qualities;
  if (Array.isArray(qualities)) {
    for (const item of qualities) {
      if (item && typeof item === "object" && typeof (item as { url?: string }).url === "string") {
        urls.push((item as { url: string }).url);
      }
    }
  }
  return urls;
}

export function validateControlsSnapshot(
  snapshot: VideoControlsSemanticSnapshot,
  lastRevision: number
): { ok: true } | { ok: false; message: string } {
  if (snapshot.revision <= lastRevision) {
    return { ok: false, message: "controls snapshot revision must increase" };
  }
  const seen = new Set<string>();
  for (const control of snapshot.controls) {
    if (!control.controlId || seen.has(control.controlId)) {
      return { ok: false, message: "controlId must be unique and non-empty" };
    }
    seen.add(control.controlId);
    if (control.role === "slider") {
      if (![control.min, control.max, control.value].every(Number.isFinite)) {
        return { ok: false, message: "slider descriptor values must be finite" };
      }
      if (control.min > control.value || control.value > control.max) {
        return { ok: false, message: "slider descriptor value is out of range" };
      }
    }
  }
  return { ok: true };
}

File: crates/lingxia-lxapp/src/inline_native/video.rs
#![allow(clippy::result_large_err)]

use super::apply::RootRegistry;
use super::types::{ErrorScope, NativeError, NativeErrorCode, NodeRef, RootLifecycle};
use serde::{Deserialize, Serialize};

const MAX_QUEUED_COMMANDS: usize = 16;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "name", rename_all = "camelCase")]
pub enum VideoCommand {
    Play,
    Pause,
    Stop,
    Seek { seconds: f64 },
    EnterFullscreen,
    ExitFullscreen,
    SetStreamSource { options: serde_json::Value },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoCommandRequest {
    pub action: String,
    pub owner: NodeRef,
    pub request_id: String,
    pub command: VideoCommand,
}

#[derive(Debug, Clone, PartialEq)]
pub enum VideoCommandOutcome {
    Applied { request_id: String },
    Queued { request_id: String },
    Rejected(Box<NativeError>),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoControlDescriptor {
    pub control_id: String,
    pub label: String,
    pub role: String,
    #[serde(default)]
    pub visible: bool,
    #[serde(default)]
    pub disabled: bool,
    #[serde(default)]
    pub value: Option<f64>,
    #[serde(default)]
    pub min: Option<f64>,
    #[serde(default)]
    pub max: Option<f64>,
    #[serde(default)]
    pub actions: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoControlsSemanticSnapshot {
    pub action: String,
    pub owner: NodeRef,
    pub revision: u64,
    pub controls: Vec<VideoControlDescriptor>,
}

#[derive(Debug, Default, Clone)]
pub struct VideoCommandQueue {
    items: Vec<VideoCommandRequest>,
}

impl VideoCommandQueue {
    pub fn len(&self) -> usize {
        self.items.len()
    }

    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }
}

pub fn apply_video_command(
    registry: &RootRegistry,
    queue: &mut VideoCommandQueue,
    request: VideoCommandRequest,
) -> VideoCommandOutcome {
    let Some(state) = registry.get(&request.owner.root()) else {
        return VideoCommandOutcome::Rejected(Box::new(command_error(
            &request,
            NativeErrorCode::RootDestroyed,
            "video command owner root is not mounted",
        )));
    };
    match state.lifecycle {
        RootLifecycle::Destroyed | RootLifecycle::Failed | RootLifecycle::Unavailable => {
            return VideoCommandOutcome::Rejected(Box::new(command_error(
                &request,
                if state.lifecycle == RootLifecycle::Destroyed {
                    NativeErrorCode::RootDestroyed
                } else {
                    NativeErrorCode::CommandFailed
                },
                "video command rejected because the root is not runnable",
            )));
        }
        RootLifecycle::Negotiating | RootLifecycle::Mounting => {
            if queue.items.len() >= MAX_QUEUED_COMMANDS {
                return VideoCommandOutcome::Rejected(Box::new(command_error(
                    &request,
                    NativeErrorCode::CommandFailed,
                    "video command queue overflow; newest command rejected",
                )));
            }
            queue.items.push(request.clone());
            return VideoCommandOutcome::Queued {
                request_id: request.request_id,
            };
        }
        RootLifecycle::Ready => {}
    }
    if !state.nodes.contains_key(&request.owner.node_key) {
        return VideoCommandOutcome::Rejected(Box::new(command_error(
            &request,
            NativeErrorCode::CommandFailed,
            "video command owner node is not in the applied tree",
        )));
    }
    let Some(node) = state.nodes.get(&request.owner.node_key) else {
        return VideoCommandOutcome::Rejected(Box::new(command_error(
            &request,
            NativeErrorCode::CommandFailed,
            "video command owner node is not in the applied tree",
        )));
    };
    if node.kind != "video" || node.node_ref.node_epoch != request.owner.node_epoch {
        return VideoCommandOutcome::Rejected(Box::new(command_error(
            &request,
            NativeErrorCode::CommandFailed,
            "video command identity does not match a video node",
        )));
    }
    if let VideoCommand::Seek { seconds } = request.command
        && !seconds.is_finite()
    {
        return VideoCommandOutcome::Rejected(Box::new(command_error(
            &request,
            NativeErrorCode::InvalidProps,
            "seek seconds must be finite",
        )));
    }
    VideoCommandOutcome::Applied {
        request_id: request.request_id,
    }
}

pub fn apply_video_controls_snapshot(
    last_revision: u64,
    snapshot: &VideoControlsSemanticSnapshot,
) -> Result<u64, NativeError> {
    if snapshot.revision <= last_revision {
        return Err(NativeError {
            code: NativeErrorCode::InvalidProps,
            scope: ErrorScope::Node,
            recoverable: true,
            root: snapshot.owner.root(),
            node: Some(snapshot.owner.clone()),
            message: "controls snapshot revision must increase".into(),
        });
    }
    let mut seen = std::collections::HashSet::new();
    for control in &snapshot.controls {
        if control.control_id.is_empty() || !seen.insert(control.control_id.clone()) {
            return Err(NativeError {
                code: NativeErrorCode::InvalidProps,
                scope: ErrorScope::Node,
                recoverable: true,
                root: snapshot.owner.root(),
                node: Some(snapshot.owner.clone()),
                message: "controlId must be unique and non-empty".into(),
            });
        }
        if control.role == "slider" {
            let (Some(min), Some(max), Some(value)) = (control.min, control.max, control.value)
            else {
                return Err(NativeError {
                    code: NativeErrorCode::InvalidProps,
                    scope: ErrorScope::Node,
                    recoverable: true,
                    root: snapshot.owner.root(),
                    node: Some(snapshot.owner.clone()),
                    message: "slider descriptor requires finite min/max/value".into(),
                });
            };
            if !min.is_finite()
                || !max.is_finite()
                || !value.is_finite()
                || min > value
                || value > max
            {
                return Err(NativeError {
                    code: NativeErrorCode::InvalidProps,
                    scope: ErrorScope::Node,
                    recoverable: true,
                    root: snapshot.owner.root(),
                    node: Some(snapshot.owner.clone()),
                    message: "slider descriptor value is out of range".into(),
                });
            }
        }
    }
    Ok(snapshot.revision)
}

fn command_error(
    request: &VideoCommandRequest,
    code: NativeErrorCode,
    message: &str,
) -> NativeError {
    NativeError {
        code,
        scope: ErrorScope::Node,
        recoverable: true,
        root: request.owner.root(),
        node: Some(request.owner.clone()),
        message: message.to_string(),
    }
}
